flink-tests/src/test/java/org/apache/flink/test/checkpointing/ChangelogPeriodicMaterializationITCase.java

```diff
@@ -0,0 +1,734 @@
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.RunnableFuture;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BooleanSupplier;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.apache.flink.shaded.guava30.com.google.common.collect.Iterables.get;
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertSame;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+/**
+ * This verifies that checkpointing works correctly for Changelog state backend with materialized
+ * state / non-materialized state.
+ */
+@RunWith(Parameterized.class)
+public class ChangelogPeriodicMaterializationITCase extends TestLogger {
+
+    private static final int TOTAL_ELEMENTS = 20;
+    private static final int NUM_TASK_MANAGERS = 1;
+    private static final int NUM_TASK_SLOTS = 4;
+    private static final int NUM_SLOTS = NUM_TASK_MANAGERS * NUM_TASK_SLOTS;
+
+    private final AbstractStateBackend delegatedStateBackend;
+
+    private StreamExecutionEnvironment env;
+
+    private URI checkpointURI;
+
+    private MiniClusterWithClientResource cluster;
+
+    @ClassRule public static final TemporaryFolder TEMPORARY_FOLDER = new TemporaryFolder();
+
+    @Parameterized.Parameters(name = "delegated state backend type ={0}")
+    public static Collection<AbstractStateBackend> parameter() {
+        return Arrays.asList(
+                new HashMapStateBackend(),
+                new EmbeddedRocksDBStateBackend(),
+                new EmbeddedRocksDBStateBackend(true));
+    }
+
+    public ChangelogPeriodicMaterializationITCase(AbstractStateBackend delegatedStateBackend) {
+        this.delegatedStateBackend = delegatedStateBackend;
+    }
+
+    @Before
+    public void setup() throws Exception {
+        cluster =
+                new MiniClusterWithClientResource(
+                        new MiniClusterResourceConfiguration.Builder()
+                                .setConfiguration(configureDSTL())
+                                .setNumberTaskManagers(NUM_TASK_MANAGERS)
+                                .setNumberSlotsPerTaskManager(NUM_TASK_SLOTS)
+                                .build());
+        cluster.before();
+        env = StreamExecutionEnvironment.getExecutionEnvironment();
```

Comment:
A minor suggestion. Since we could have different `env` for different test cases, we would not make the environment as a whole private field here but leave it in each test case, and pass it into `buildStreamGraph` and `waitUntilJobFinished`.